src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs

```diff
@@ -605,6 +623,28 @@ internal override void DecodeWellKnownAttribute(ref DecodeWellKnownAttributeArgu
                 // NullableAttribute should not be set explicitly.
                 arguments.Diagnostics.Add(ErrorCode.ERR_ExplicitNullableAttribute, arguments.AttributeSyntaxOpt.Location);
             }
+            else if (attribute.IsTargetAttribute(this, AttributeDescription.NotNullWhenFalseAttribute))
+            {
+                if (this.ContainingSymbol.GetTypeOrReturnType().SpecialType == SpecialType.System_Boolean)
+                {
+                    arguments.GetOrCreateData<CommonParameterWellKnownAttributeData>().HasNotNullWhenFalseAttribute = true;
+                }
+                else
+                {
+                    arguments.Diagnostics.Add(ErrorCode.ERR_AttributeRequiresBoolReturn, arguments.AttributeSyntaxOpt.Location, AttributeDescription.NotNullWhenFalseAttribute.Name);
+                }
+            }
+            else if (attribute.IsTargetAttribute(this, AttributeDescription.EnsuresNotNullAttribute))
+            {
+                if (this.Type.IsValueType)
```

Comment:
>this.Type.IsValueType [](start = 20, length = 21)

What if type is an unconstraint generic type parameter?

Reply:
Unconstrained generic type gives `IsValueType` as `false` and no warning is produced.
That said, I'm removing those warnings in follow-up PR, from discussion with Neal and Chuck (those warning create compat issue once the attributes are released, if someone uses the attribute with C# 7 compiler then upgrades to C# 8 compiler).